### ISSUE TITLE
"Unable to set up keys" with Active Directory User

### ISSUE THREAD

**Issue by Nyxtorm** [matrix-org/matrix-synapse-ldap3#178]
Hi,

When I use the LDAP module, I can authenticate myself without any problem and send messages to another user but I constantly have an error concerning the creation of the keys.

If I use a local user created via `register_new_matrix_user`, I don't encounter the problem, which leads me to a problem with the LDAP/Active Directory module.

### Error on Element web : 

![](https://i.imgur.com/IVZ7QAG.png)

### Browser console : 

![](https://i.imgur.com/tyOZz3M.png)

### Synapse logs : 

```
2023-04-04 18:17:52,820 - ldap_auth_provider - 178 - INFO - POST-130 - User authenticated against LDAP server: ldap://ad:389 - cleartext - user: CN=XXXXXX XXXXXX,OU=Users,OU=XXX,DC=XXXXXX,DC=LAN - not lazy - unbound - closed - <no socket> - tls not started - not listening - SyncStrategy - internal decoder
2023-04-04 18:17:52,826 - synapse.handlers.auth - 591 - INFO - POST-130 - Auth completed with creds: {'m.login.password': '@XXXXXX.XXXXXX:XX.XXXXXX.fr'}. Client dict has keys: ['master_key', 'self_signing_key', 'user_signing_key']
2023-04-04 18:17:52,827 - synapse.http.server - 107 - INFO - POST-130 - <XForwardedForRequest at 0xYYYYYYYYY method='POST' uri='/_matrix/client/unstable/keys/device_signing/upload' clientproto='HTTP/1.1' site='8008'> SynapseError: 403 - Invalid auth
2023-04-04 18:17:52,827 - synapse.access.http.8008 - 460 - INFO - POST-130 - 10.10.10.10 - 8008 - {@XXXXXX.XXXXXX/XXXXXX.lan:XXX.XXXXXX.fr} Processed request: 0.358sec/0.000sec (0.004sec, 0.000sec) (0.001sec/0.015sec/8) 48B 403 "POST /_matrix/client/unstable/keys/device_signing/upload HTTP/1.1" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/111.0.0.0 Safari/537.36" [0 dbevts]
```

### Config for LDAP module : 

```
modules:
 - module: "ldap_auth_provider.LdapAuthProviderModule"
   config:
     enabled: true
     mode: "search"
     uri: "ldap://ad:389"
     base: "DC=XXXXXXXXXXXXX,DC=LAN"
     active_directory: true
     default_domain: XXXXXXXXXXXXX.LAN
     attributes:
        uid: "userPrincipalName"
        mail: "mail"
        name: "cn"
     bind_dn: "CN=XXXXXX,OU=Users,OU=XXX,DC=XXXXXXXXXXXXX,DC=LAN"
     bind_password: "xxxxxxxxxxxxx"
```

I can't understand why I get this 403 error with an AD user but not with a regular user (I'm able to generate keys without any issues). I can't get more information, it looks like a bug but I can't find any similar case.

Thanks for your help !

**Comment by Nyxtorm** [comment]
I tried today with Cinny desktop app, the problem is the same. So the error must not come from Element.

**Comment by Nyxtorm** [comment]
In the meantime, I was able to find the source of the problem and work around it (without fixing it because I'm not expert enough).

The key generation problem seems to be related to the fact that I was using the `default_domain` option for LDAP, and authenticating with `john.doe` instead of `john.doe/mydomain.lan`.

Once I removed the `default_domain` option and authenticated with the full username, there was no problem with key generation.

This is not a fix but hopefully it can help someone in the same situation!

Sorry for my English.
